@here/harp-mapview/lib/geometry/SolidLineMesh.ts

```diff
@@ -0,0 +1,451 @@
+        ).getBoundingSphere(sphere);
+    }
+
+    // If geometry is not displaced, the bounding sphere is placed in the center of the computed
+    // bounding box and expanded until it contains all vertices.
+    const center = bbox.getCenter(sphere.center);
+    let maxRSq = 0;
+    for (let i = featureBeginIndex; i < featureEndIndex; i += SEGMENT_STRIDE) {
+        maxRSq = Math.max(
+            maxRSq,
+            center.distanceToSquared(vertex.fromBufferAttribute(pos, indices[i])),
+            center.distanceToSquared(vertex.fromBufferAttribute(pos, indices[i + VERTEX_STRIDE]))
+        );
+    }
+    sphere.radius = Math.sqrt(maxRSq);
+    return sphere;
+}
+
+/**
+ * Finds the intersection of a ray with a extruded line.
+ * @param ray Intersection ray in object local space.
+ * @param line The centerline.
+ * @param vExtrusion Line extrusion vector.
+ * @param normal Extrusion plane normal.
+ * @param hWidth Extrusion half width.
+ * @returns Distance of the extruded line intersection to the ray origin.
+ */
+function intersectExtrudedLine(
+    ray: THREE.Ray,
+    line: THREE.Line3,
+    vExtrusion: THREE.Vector3,
+    normal: THREE.Vector3,
+    hWidth: number
+): number {
+    const matrix = new THREE.Matrix4().makeBasis(
+        line.delta(new THREE.Vector3()).normalize(),
+        vExtrusion,
+        normal
+    );
+    const obb = new OrientedBox3(
+        line.getCenter(tmpV4),
+        matrix,
+        new THREE.Vector3(line.distance() / 2, hWidth, hWidth)
+    );
+    if (obb.contains(ray.origin)) {
+        return 0;
+    }
+    return obb.intersectsRay(ray) ?? Infinity;
+}
+
+/**
+ * Finds the intersection of a ray with the closest end cap of a extruded line.
+ * @param ray Intersection ray in object local space.
+ * @param line The centerline.
+ * @param hWidth Extrusion half width.
+ * @returns Distance of the end cap intersection to the ray origin.
+ */
+function intersectClosestEndCap(ray: THREE.Ray, line: THREE.Line3, hWidth: number): number {
+    const sphere = new THREE.Sphere(line.start, hWidth);
+    const startCapT = sphere.containsPoint(ray.origin)
+        ? 0
+        : ray.intersectSphere(sphere, tmpV4)
+        ? tmpV4.sub(ray.origin).length()
+        : Infinity;
+    sphere.center.copy(line.end);
+    const endCapT = sphere.containsPoint(ray.origin)
+        ? 0
+        : ray.intersectSphere(sphere, tmpV4)
+        ? tmpV4.sub(ray.origin).length()
+        : Infinity;
+    return Math.min(startCapT, endCapT);
+}
+
+/**
+ * Intersects line
+ * @param ray Intersection ray in object local space.
+ * @param line The line to intersect.
+ * @param vExtrusion Line extrusion vector.
+ * @param hWidth The line's extrusion half width.
+ * @param hWidthSq The line's extrusion half width squared.
+ * @param plane The extrusion plane.
+ * @param interPlane The intersection of the ray with the extrusion plane.
+ * @param interLine The ray intersetion with the extruded line.
```

Comment:
Please explain are they input/output params. How can I distinguish which of the params is actually computed?

Reply:
done